pgrx/src/varlena.rs

```diff
@@ -21,7 +21,10 @@ pub unsafe fn set_varsize_4b(ptr: *mut pg_sys::varlena, len: i32) {
     // 	(((varattrib_4b *) (PTR))->va_4byte.va_header = (((uint32) (len)) << 2))
 
     // SAFETY:  A varlena can be safely cast to a varattrib_4b
-    (*ptr.cast::<pg_sys::varattrib_4b>()).va_4byte.as_mut().va_header = (len as u32) << 2;
+    let header = &mut (*ptr.cast::<pg_sys::varattrib_4b>()).va_4byte.deref_mut().va_header;
```

Comment:
Ah, I see, it needs ManuallyDrop because Rust doesn't require `!Drop` to put something in a union, but instead requires `Copy`, and we can't really impl `Copy` for a dynamically-sized length... got it. Yeah, that makes sense. And then we get this weird special-case of auto-DerefMut being disabled, so we have to call it explicitly.

Okay, my comment about `core::ptr::addr_of_mut!` remains the same here, if you can get that place-expression to work instead, , but otherwise this is fine.

...wait, this is open-coded in `impl<'a> IntoDatum for &'a [u8]`.

...that should be replaced by this, and the `try_into` should be turned into a debug assertion.

Reply:
This should be resolved now - `IntoDatum`'s separate implementation of this behavior has been replaced with `set_varsize_4b()`, and there's no longer an invocation of the zero-sized-type's `.as_mut_ptr()`. Let me know if it still does not look correct.